util/graph.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package util
+
+// DFSTraversal defines the basic interface required to perform a depth
+// first traveral.
+type DFSTraversal interface {
+	Edges(T) []T
+	Equals(T, T) bool
+	Visited(T) bool
+}
+
+// DFS returns a path from node a to node z found by performing
+// a depth first traversal. If no path is found, an empty slice is returned.
+func DFS(t DFSTraversal, a, z T) []T {
+	p := dfsRecursive(t, a, z, []T{})
+	for i := len(p)/2 - 1; i >= 0; i-- {
+		o := len(p) - i - 1
+		p[i], p[o] = p[o], p[i]
+	}
+	return p
+}
+
+func dfsRecursive(t DFSTraversal, u, z T, path []T) []T {
+	if t.Visited(u) {
```

Comment:
Does this alg work?  I thought we needed to run dfs on every node in the graph in case some nodes aren't reachable from others.  And because of that we need to keep start/end times on each node (instead of just the boolean Visited) so that when we encounter a new node we can differentiate (i) the node has been started but not finished (a cycle) from (ii) a node that was started and finished but for which there is no path from the new node back to the node we came from.